### PR TITLE
RDF to Object conversion (closing #555)

### PR DESCRIPTION
**Pull request by anatoly-scherbakov**
# Why

Trying to solve #555.

# What

Clarify what we do when can't serialize a number from RDF into JSON.

***
<a href="https://pr-preview.s3.amazonaws.com/iolanta-tech/json-ld-api/pull/619.html" title="Last updated on Nov 29, 2024, 7:39 PM UTC (a6132a4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-api/619/a24a4b3...iolanta-tech:a6132a4.html" title="Last updated on Nov 29, 2024, 7:39 PM UTC (a6132a4)">Diff</a>

### CODE REVIEW

**Comment by anatoly-scherbakov** [comment]
I don't think I can invite reviewers to this PR :thinking: @pchampin @gkellogg

**Comment by gkellogg** [comment]
> I don't think I can invite reviewers to this PR 🤔 @pchampin @gkellogg

I'll look at your permissions, you should have full permissions on this and other repos.

File: index.html [hunk 1]
```diff
@@ -5156,17 +5156,38 @@ <h3>Overview</h3>
...
         and <a>blank node identifiers</a>.
+
         If the <a data-link-for="JsonLdOptions">useNativeTypes</a> flag is set to <code>true</code>,
         <a>RDF literals</a> with a
         <a>datatype IRI</a>
-        that equals <code>xsd:integer</code> or <code>xsd:double</code> are converted
-        to a <a>JSON numbers</a> and <a>RDF literals</a>
+        that equals
+
+        <ul>
```
Comment (gkellogg) [inline index.html:5165]: The paragraph will need to be broken up, as a `<p>` can not include a `<ul>`.

File: index.html [hunk 2]
```diff
@@ -5156,17 +5156,38 @@ <h3>Overview</h3>
...
+        to <a>JSON numbers</a>,
+            <ul>
+              <li>
+                unless such a conversion is impossible, for instance, in cases like
+                <code>"+INF"^^xsd:double</code>, in which instance <a data-link-for="JsonLdOptions">useNativeTypes</a> flag is ignored,
+                and the literal value is kept in string form,
+                in accordance with <a data-cite="RFC8785#section-appendix.d">RFC 8785, Appendix D. Dealing with Big Numbers</a>.
+              </li>
+            </ul>
+          </li>
```
Comment (gkellogg) [inline index.html:5177]: This should also account for the other cases called out in JCS about numbers that are too large, or carry too much precision, to be represented by either xsd:integer or xsd:double. For example, `1.4e+9999`, and `9223372036854775807`.

**Review by gkellogg — CHANGES_REQUESTED** [review]
The changes are to the informative summary, but the issue actually calls out 2.4.3 (was 2.3.4) which is normative; that's the important area to address.

**Comment by gkellogg** [comment]
@anatoly-scherbakov, if you can create branches at json-ld-api rather than on your own fork, it is easier for others to collaborate.

**Comment by gkellogg** [comment]
As an alternative, we may treat these as error conditions, which is what JCS does.

> [3.2.2.3. ](https://www.rfc-editor.org/rfc/rfc8785#section-3.2.2.3)[Serialization of Numbers](https://www.rfc-editor.org/rfc/rfc8785#name-serialization-of-numbers)
> ...
> Note: Since Not a Number (NaN) and Infinity are not permitted in JSON, occurrences of NaN or Infinity MUST cause a compliant JCS implementation to terminate with an appropriate error.

**Comment by davidlehn** [comment]
Can you please replace the emoji in the commit titles with words.

File: index.html [hunk 3]
```diff
@@ -5156,17 +5156,38 @@ <h3>Overview</h3>
         and generating a JSON-LD document in expanded form for all
         <a>RDF literals</a>, <a>IRIs</a>
         and <a>blank node identifiers</a>.
+
         If the <a data-link-for="JsonLdOptions">useNativeTypes</a> flag is set to <code>true</code>,
         <a>RDF literals</a> with a
         <a>datatype IRI</a>
-        that equals <code>xsd:integer</code> or <code>xsd:double</code> are converted
-        to a <a>JSON numbers</a> and <a>RDF literals</a>
+        that equals
```
Comment (davidlehn) [inline index.html:5163]: This list markup inside a sentence style is a bit difficult to read.  I'd suggest rewording the changes and using a similar style to what is elsewhere in the spec.

File: index.html [hunk 4]
```diff
@@ -5156,17 +5156,89 @@ <h3>Overview</h3>
...
-        as described in
-        <a class="sectionRef" href="#data-round-tripping"></a>.
+
+        If the <a data-link-for="JsonLdOptions">useNativeTypes</a> flag is set
+        to <code>true</code>, <a>RDF literals</a> are converted to JSON values
+        based on their <a>datatype IRI</a> and <a>lexical form</a> as described
+        in <a class="sectionRef" href="#data-round-tripping"></a>.
+      </p>
+
+      <table class="simple">
```
Comment (gkellogg) [inline index.html:5166]: The table formatting needs some work; probably vertical alignment and borders to make it understandable.

**Review by gkellogg — CHANGES_REQUESTED** [review]

File: index.html (hunk 4 above)
Comment (anatoly-scherbakov) [reply index.html:5166]: Yes, this is a rather draft version; I'd like it to look like `<table class="simple">` in YAML-LD spec but I haven't looked yet into what CSS markup is needed for that.

Comment (anatoly-scherbakov) [reply index.html:5166]: …ReSpec has changed table formatting: https://github.com/speced/respec/issues/4667

A pity, I liked the previous blue style.

Comment (anatoly-scherbakov) [reply index.html:5208]: image

Updated.

**Review by gkellogg — CHANGES_REQUESTED** [review]
Minor change to table head:  th elements need to be wrapped inside a tr.

File: index.html (hunk 4 above)
Comment (anatoly-scherbakov) [reply index.html:5208]: Yes, that's what I did to obtain the formatting on the screenshot above.

File: index.html [hunk 5]
```diff
@@ -5156,17 +5198,91 @@ <h3>Overview</h3>
...
-        with a <a>datatype IRI</a>
-        that equals <code>xsd:boolean</code> are converted to <code>true</code> or
-        <code>false</code> based on their
-        <a>lexical form</a>
-        as described in
-        <a class="sectionRef" href="#data-round-tripping"></a>.
+
+        If the <a data-link-for="JsonLdOptions">useNativeTypes</a> flag is set
+        to <code>true</code>, <a>RDF literals</a> are converted to JSON values
+        based on their <a>datatype IRI</a> and <a>lexical form</a> as described
```
Comment (TallTed) [inline index.html:5204]:
```suggestion
        based on their <a>datatype IRI</a> and <a>lexical form</a>, as described
```

File: index.html [hunk 6]
```diff
@@ -5477,10 +5593,24 @@ <h3>Algorithm</h3>
...
+                  according to [[XMLSCHEMA11-2]], and if it is a numeric value
+                  conformant to [[IEEE-754-2008]], then set <var>converted value</var>
                   to the result of converting the
                   <a>lexical form</a>
-                  to a JSON <a>number</a>.</li>
+                  to a JSON <a>number</a>.
+                </li>
+                <li>
+                  Otherwise, if <a>lexical form</a> expresses a numerical value
+                  outside of the range mandated by [[IEEE-754-2008]], then set
```
Comment (TallTed) [inline index.html:5604]: There are several ranges supported by [IEE-754-2008 (and IEE-754-2019, the latest rev, to which we might want to update our references)](https://en.wikipedia.org/wiki/IEEE_754), depending on the subtype of the numeric.
```suggestion
                  Otherwise, if <a>lexical form</a> expresses a numeric value
                  outside of the ranges supported by [[IEEE-754-2008]], then set
```

File: tests/fromRdf-manifest.jsonld [hunk 7]
```diff
@@ -195,6 +195,26 @@
       "purpose": "Check list generation with rdf:first property and rdf:nil value.",
       "input": "fromRdf/0026-in.nq",
       "expect": "fromRdf/0026-out.jsonld"
+    }, {
+      "@id": "#t0027",
+      "@type": ["jld:PositiveEvaluationTest", "jld:FromRDFTest"],
+      "name": "native types flag set to true is ignored if the number provided is invalid in JSON",
+      "purpose": "Literals with datatype xsd:boolean, xsd:integer, and xsd:double are serialized using native scalar values if possible",
```
Comment (TallTed) [inline tests/fromRdf-manifest.jsonld:202]:
```suggestion
      "purpose": "Literals with datatype xsd:boolean, xsd:integer, and xsd:double are serialized using native scalar values, if possible",
```

**Review by gkellogg — APPROVED** [review]
Looks good, subject to @TallTed's suggestions. We may want to revisit what happens if `Infinity` or `NaN` is attempted to be serialized.

**Comment by anatoly-scherbakov** [comment]
@TallTed created a separate issue about the version of the IEEE754 standard we refer to: https://github.com/w3c/json-ld-api/issues/620

**Comment by anatoly-scherbakov** [comment]
@gkellogg do you mean we could convert them to JSON Strings instead of raising a JSON-LD processing error?

**Comment by anatoly-scherbakov** [comment]
GitHub tells me that

> You're not authorized to push to this branch. Visit https://docs.github.com/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches for more information.

Not sure why though, it might be related to the fact that I opened it in my fork.

File: index.html [hunk 8]
```diff
@@ -5156,17 +5198,91 @@ <h3>Overview</h3>
...
+
+          <tr>
+            <td>Whole number out of that range</td>
+            <td>JSON string literal</td>
+          </tr>
+
+          <tr>
+            <td>Any other value</td>
+            <td class="error">Throw <a data-link-for="JsonLdErrorCode">invalid JSON literal</a></td>
+          </tr>
```
Comment (pchampin) [inline index.html:5240]: I don't think that is a good idea, but I'll comment on that in the algorithm itself, below.

File: index.html [hunk 9]
```diff
@@ -5156,17 +5198,91 @@ <h3>Overview</h3>
...
+              Non-numeric values:
+
+              <ul>
+                <li><code>"NaN"^^xsd:double</code>,</li>
+                <li><code>"+INF"^^xsd:double</code>,</li>
+                <li><code>"-INF"^^xsd:double</code></li>
+              </ul>
+            </td>
+            <td class="error" rowspan="2">Throw <a data-link-for="JsonLdErrorCode">invalid JSON literal</a></td>
+          </tr>
```
Comment (pchampin) [inline index.html:5264]: Same as above; I don't think that is a good idea, see my comments in the algorithm.

**Review by pchampin — CHANGES_REQUESTED** [review]
See my comments in the "Files changed" view.

**Review by gkellogg — CHANGES_REQUESTED** [review]
I question if this is a class-2 editorial change; I think it's class-3 change, which will require Candidate Correction marking. We can discuss.

**Comment by pchampin** [comment]
> I question if this is a class-2 editorial change; I think it's class-3 change, which will require Candidate Correction marking. We can discuss.

+1, this falls under "clears up an ambiguity or under-specified part of the specification" which is part of "class 3".
+1 also to discuss

File: index.html [hunk 10]
```diff
@@ -5156,17 +5198,89 @@ <h3>Overview</h3>
...
+
+          <tr>
+            <th rowspan="3"><code>xsd:integer</code></th>
+            <td>Whole number in the <code>[-2<sup>53</sup> + 1, 2<sup>53</sup> + 1]</code> range, representable as per [[IEEE-754-2008]]</td>
+            <td><a>JSON Number</a></td>
+          </tr>
+
+          <tr>
+            <td>Whole number out of that range</td>
+            <td>JSON string literal</td>
```
Comment (gkellogg) [inline index.html:5232]: Actually, this needs to be serialized as a value object with `@type`. The integer `1e54` (`1000000000000000078291540404596243842305360299886116864`) is outside of that range, but the value object needs to include `"@type": "xsd:integer".

File: index.html [hunk 11]
```diff
@@ -5156,17 +5198,89 @@ <h3>Overview</h3>
...
+        <tbody>
+          <tr>
+            <th><code>xsd:string</code></th>
+            <td>An RDF string literal</td>
+            <td>JSON string literal</td>
+          </tr>
+
+          <tr>
+            <th rowspan="3"><code>xsd:integer</code></th>
+            <td>Whole number in the <code>[-2<sup>53</sup> + 1, 2<sup>53</sup> + 1]</code> range, representable as per [[IEEE-754-2008]]</td>
```
Comment (gkellogg) [inline index.html:5226]: There are more constraints in the [lexical form of an xsd:integer](https://www.w3.org/TR/xmlschema11-2/#integer-lexical-representation) than this.

Also, while it's likely consistent with JCS/ECMAScript, such numbers also need to be representable as integers in JCS/ECMAScript. See [Appendix B](https://www.rfc-editor.org/rfc/rfc8785#json.ieee754.test). 
We should be careful to not mis-describe this, and just refer to XSD Datatypes 1.1 and JCS.

File: index.html [hunk 12]
```diff
@@ -5156,17 +5198,89 @@ <h3>Overview</h3>
...
+
+          <tr>
+            <th rowspan="4"><code>xsd:double</code></th>
+            <td>Real number in the <code>[-2<sup>53</sup> + 1, 2<sup>53</sup> + 1]</code> range, representable as per [[IEEE-754-2008]]</td>
+            <td><a>JSON Number</a></td>
+          </tr>
+
+          <tr>
+            <td>Real number out of that range</td>
+            <td>JSON string literal</td>
```
Comment (gkellogg) [inline index.html:5248]: See above comment; the ranges should be consistent with XSD and JCS and probably shouldn't re-define the criteria.

File: index.html [hunk 13]
```diff
@@ -5477,10 +5593,24 @@ <h3>Algorithm</h3>
...
+                  outside of the ranges supported by [[IEEE-754-2008]], then set
+                  <var>converted value</var> to a JSON string literal containing
+                  said lexical form.
+                </li>
+                <li>
+                  Otherwise, if <a>lexical form</a> expresses a non-numeric value,
+                  such as <code>"+INF"^^xsd:double</code>, an
+                  <a data-link-for="JsonLdErrorCode">invalid JSON literal</a>
+                  error has been detected and processing is aborted.
+                </li>
```
[the remark this answers is not in the document]
Comment (gkellogg) [reply index.html:5611]: Any specific range examples should be careful to be informative, to not get ahead of XSD. Also, it will need to be consistent with JCS.

It might be that they way to do this is to just use JCS for serializing strings, and require that the result of serializing to JCS must result that is in the value space of xsd:integer or xsd:double, respectively.

File: tests/fromRdf-manifest.jsonld [hunk 14]
```diff
@@ -195,6 +195,26 @@
...
+    }, {
+      "@id": "#t0028",
+      "@type": ["jld:NegativeEvaluationTest", "jld:FromRDFTest"],
+      "name": "native types flag set to true is ignored if the number provided is invalid in JSON",
+      "purpose": "+INF^^xsd:double causes a failure to serialize in JSON.",
+      "option": {
+        "useNativeTypes": true
+      },
+      "input": "fromRdf/0028-in.nq",
+      "expectErrorCode": "invalid JSON literal"
```
Comment (gkellogg) [inline tests/fromRdf-manifest.jsonld:217]: This is not invalid, it just can't be serialized as a native value.

**Review by gkellogg — CHANGES_REQUESTED** [review]
I agree with @pchampin, but we need to be clear that it needs to be valid both as xsd:double and JCS number, and fallback to the expanded representation.

**Comment by anatoly-scherbakov** [comment]
I believe I was misunderstanding a few things when writing this PR, sorry guys. Instead of reverting much of these changes, I just created another PR: https://github.com/w3c/json-ld-api/pull/625

And I am closing this one; it won't disappear so as to maintain the history, just will be closed.
